[WebXR] Refactor the layer backings to allow WebGL layers

svillar opened the pull request:
#### d57237a9c78977bfa95cbce9804daf82cfd42131
<pre>
[WebXR] Refactor the layer backings to allow WebGL layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=302653">https://bugs.webkit.org/show_bug.cgi?id=302653</a>

Reviewed by Mike Wyrzykowski.

WebXR layers API can be implemented using either a WebGL backend (as in
the original specification) or using the WebGPU bindings (as it&apos;s
currently implemented for Cocoa platforms). The current code however was
developed as if only the WebGPU option was possible and thus does not
easily allow the co-existence of a WebGL backend. That&apos;s why we need to
refactor the class that implements the layers backend, known as the
&quot;backing&quot;.

A new superclass called XRLayerBacking was added. That&apos;s the class that
would be used by the XR composition layers. The WebGPU implementation of
the backing know inherits from that new class.

Apart from that we&apos;re also moving the reference to the backing to the
XRCompositionLayer as that would be needed for all composition layer
types and not only for the projection layer.

No new tests required, we&apos;re &quot;just&quot; refactoring some code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRLayerBacking.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRProjectionLayer.h:
* Source/WebCore/Modules/webxr/XRCompositionLayer.cpp:
(WebCore::XRCompositionLayer::XRCompositionLayer):
(WebCore::XRCompositionLayer::backing):
* Source/WebCore/Modules/webxr/XRCompositionLayer.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::getSubImage):
* Source/WebCore/Modules/webxr/XRLayerBacking.h:
(WebCore::XRLayerBacking::startFrame):
(WebCore::XRLayerBacking::endFrame):
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::XRProjectionLayer):
(WebCore::XRProjectionLayer::endFrame):
(WebCore::XRProjectionLayer::textureWidth const):
(WebCore::XRProjectionLayer::textureHeight const):
(WebCore::XRProjectionLayer::textureArrayLength const):
(WebCore::XRProjectionLayer::backing):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:
(WebCore::XRProjectionLayer::create):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/303525@main">https://commits.webkit.org/303525@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/163be2a636f2ee8e10248c6a5a47760662734548

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5148 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140177 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84680 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5431 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101416 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135599 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118834 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82209 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1412 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83413 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112667 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36949 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142834 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4818 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37538 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109789 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4900 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109966 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27885 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3672 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115106 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58277 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4872 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33454 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4708 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68323 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
looks great - thank you!

svillar committed:
[WebXR] Refactor the layer backings to allow WebGL layers
https://bugs.webkit.org/show_bug.cgi?id=302653

Reviewed by Mike Wyrzykowski.

WebXR layers API can be implemented using either a WebGL backend (as in
the original specification) or using the WebGPU bindings (as it's
currently implemented for Cocoa platforms). The current code however was
developed as if only the WebGPU option was possible and thus does not
easily allow the co-existence of a WebGL backend. That's why we need to
refactor the class that implements the layers backend, known as the
"backing".

A new superclass called XRLayerBacking was added. That's the class that
would be used by the XR composition layers. The WebGPU implementation of
the backing know inherits from that new class.

Apart from that we're also moving the reference to the backing to the
XRCompositionLayer as that would be needed for all composition layer
types and not only for the projection layer.

No new tests required, we're "just" refactoring some code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRLayerBacking.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUXRProjectionLayer.h:
* Source/WebCore/Modules/webxr/XRCompositionLayer.cpp:
(WebCore::XRCompositionLayer::XRCompositionLayer):
(WebCore::XRCompositionLayer::backing):
* Source/WebCore/Modules/webxr/XRCompositionLayer.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::getSubImage):
* Source/WebCore/Modules/webxr/XRLayerBacking.h:
(WebCore::XRLayerBacking::startFrame):
(WebCore::XRLayerBacking::endFrame):
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::XRProjectionLayer):
(WebCore::XRProjectionLayer::endFrame):
(WebCore::XRProjectionLayer::textureWidth const):
(WebCore::XRProjectionLayer::textureHeight const):
(WebCore::XRProjectionLayer::textureArrayLength const):
(WebCore::XRProjectionLayer::backing):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:
(WebCore::XRProjectionLayer::create):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/303525@main
